ludwig/schema/features/augmentation/utils.py

```diff
@@ -0,0 +1,156 @@
+from dataclasses import field
+from typing import List, Union
+
+from marshmallow import fields, ValidationError
+
+from ludwig.api_annotations import DeveloperAPI
+from ludwig.constants import TYPE
+from ludwig.schema import utils as schema_utils
+from ludwig.utils.registry import Registry
+
+_augmentation_config_registry = Registry()
+
+
+@DeveloperAPI
+def get_augmentation_config_registry() -> Registry:
+    return _augmentation_config_registry
+
+
+@DeveloperAPI
+def register_augmentation_config(name: str, features: Union[str, List[str]]):
+    if isinstance(features, str):
+        features = [features]
+
+    def wrap(cls):
+        for feature in features:
+            augmentation_registry = get_augmentation_config_registry().get(feature, {})
+            augmentation_registry[name] = cls
+            get_augmentation_config_registry()[feature] = augmentation_registry
+        return cls
+
+    return wrap
+
+
+@DeveloperAPI
+def get_augmentation_cls(feature: str, name: str):
+    return get_augmentation_config_registry()[feature][name]
+
+
+@DeveloperAPI
+def get_augmentation_classes(feature: str):
+    return get_augmentation_config_registry()[feature]
+
+
+@DeveloperAPI
+def AugmentationContainerDataclassField(feature_type: str, default=[], description=""):
+    """Custom dataclass field that when used inside a dataclass will allow the user to specify an augmentation
+    config.
+
+    Args:
+        default: The default augmentation config to use.
+        description: The description of the augmentation config.
+
+    Returns: Initialized dataclass field that converts an untyped dict with params to an augmentation config.
+    """
+
+    class AugmentationContainerMarshmallowField(fields.Field):
+        """Custom marshmallow field that deserializes a list for a valid augmentation config from the
+        augmentation_registry and creates a corresponding JSON schema for external usage."""
+
+        def _deserialize(self, value, attr, data, **kwargs):
+            assert isinstance(value, list), "Augmentation config must be a list."
+
+            augmentation_list = []
+            for augmentation in value:
+                augmentation_op = augmentation[TYPE]
+                augmentation_cls = get_augmentation_cls(augmentation_op)
```

Comment:
I think you are missing the `feature_type` argument here?

Reply:
done

Reply:
Though it is curious this error was not flagged in any unit test. 

When does `_deseralize()` method invoked?  What would be a reasonable way to unit test this code path?

Reply:
I think it passed because `get_augmentation_cls` will use `image` by default if nothing else is provided, which makes sense.

`_deserialize` can only be invoked at runtime when Ludwig tries to convert a user config to a config object. If the user specifies nothing for `augmentation` then its actually skipped because the compile-time `default` will be used. Any test that attempts training with a user-specified `augmentation` will hit this code path.